cpp/dolfinx/refinement/utils.cpp

```diff
@@ -372,24 +374,42 @@ refinement::adjust_indices(const common::IndexMap& index_map, std::int32_t n)
   // Add in an extra "n" indices at the end of the current local_range
   // of "index_map", and adjust existing indices to match.
 
-  // Get number of new indices on all processes
+  // Get offset for n for this process
   MPI_Comm comm = index_map.comm();
-  int mpi_size = dolfinx::MPI::size(comm);
-  int mpi_rank = dolfinx::MPI::rank(comm);
-  std::vector<std::int32_t> recvn(mpi_size);
-  MPI_Allgather(&n, 1, MPI_INT32_T, recvn.data(), 1, MPI_INT32_T, comm);
-  std::vector<std::int64_t> global_offsets = {0};
-  for (std::int32_t r : recvn)
-    global_offsets.push_back(global_offsets.back() + r);
+  const std::int64_t num_local = n;
+  std::int64_t global_offset = 0;
+  MPI_Exscan(&num_local, &global_offset, 1, MPI_INT64_T, MPI_SUM, comm);
+
+  // Use MPI neighbors to get offsets for ghosts. Use the source of the
+  // forward comm to get ghost entry neighbors and create a dictionary
+  // to lookup from the global rank
+  MPI_Comm comm_fwd = index_map.comm(common::IndexMap::Direction::forward);
+  std::vector<int> neighbors = MPI::neighbors(comm_fwd)[0];
+  std::map<int, int> proc_to_nbr;
+  for (std::size_t i = 0; i < neighbors.size(); ++i)
+    proc_to_nbr.insert({neighbors[i], i});
 
-  std::vector global_indices = index_map.global_indices();
+  // Communicate offset to neighbors
+  std::vector<std::int64_t> neighbor_offsets(neighbors.size(), 0);
+  // Ensure allocation, in case where neighbors.size() == 0, needed for some MPI
+  neighbor_offsets.reserve(1);
+  MPI_Neighbor_allgather(&global_offset, 1, MPI_INT64_T,
+                         neighbor_offsets.data(), 1, MPI_INT64_T, comm_fwd);
 
   const std::vector<int>& ghost_owners = index_map.ghost_owner_rank();
   int local_size = index_map.size_local();
-  for (int i = 0; i < local_size; ++i)
-    global_indices[i] += global_offsets[mpi_rank];
+  std::vector<std::int64_t> global_indices = index_map.global_indices();
```

Comment:
This looks very circular going to-and-from global and neighbourhood ranks several times. I think it can be simpler and the map construction avoided.

Reply:
IndexMap should have a function that returns the `ghost_owners_neighbor_rank`